### PR TITLE
Allow CSharpSyntaxGenerator to be used as a source generator

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This takes a fairly hacky approach to updating the code by converting the CSharpSyntaxGenerator project to build twice: once as a
netcoreapp3.1 console app like it always did, and now also as a netstandard2.0 library that is a source generator. The source generator one only supports building the .cs outputs that are consumed as regular APIs, and ignores the test or grammar outputs for now.

### CODE REVIEW

File: eng/targets/Imports.targets
Comment (sharwell) [inline eng/targets/Imports.targets:174]: ❔ Can we use the new property for this?

/cc @mavasani

Comment (jasonmalinowski) [reply eng/targets/Imports.targets:174]: What new property?

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj
Comment (jasonmalinowski) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj:8]: All of the code in this project was in CSharpSyntaxGenerator; I don't know why the RootNamespace was what it was originally set to.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj
Comment (jasonmalinowski) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj:23]: This was why I'm doing this funky build hackery: if you're producing the grammar tool, it actually builds by linking in the source files of the compiler. I wanted the assurance that the source generator form was behaving as a true 'source generator' and any information about the source was actually being read out of the inputs rather than being linked at the time of build. If it was consuming stuff from the earlier build then we could end up in some really funky situations where a change would have to be made the compiler directory, then you'd have to build the new generator, and then build the other project again to get everything into sync.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (jasonmalinowski) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:66]: @chsienki Should we have an API that lets you do AddSource and directly give it a StringBuilder? Having to write this myself seems icky.

File: eng/targets/Imports.targets
Comment (mavasani) [reply eng/targets/Imports.targets:174]: @jasonmalinowski https://github.com/dotnet/roslyn/pull/46669. We should be able to get rid of `DisableAnalyzers` target and `UseRoslynAnalyzers` property in Roslyn completely in favor of just setting `RunAnalyzersDuringBuild` appropriately.

Comment (jasonmalinowski) [reply eng/targets/Imports.targets:174]: Ah, right forgot that went in. Yeah, we can try just using that now.

**Comment by 333fred** [comment]
I want to make sure that we don't merge this before we've validated the experience works on vscode, or this will effectively break my ability to work.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (sharwell) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:66]: Why not use the overload that takes a string?

```suggestion
                context.AddSource(hintName, stringBuilder.ToString());
```

**Comment by jasonmalinowski** [comment]
Marking back as draft as there's some toolset changes and other stuff that needs fixing.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (jasonmalinowski) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:66]: @sharwell: because that'll allocate a 2MB string. :smile:

**Comment by jasonmalinowski** [comment]
So I'm able to see the generator running fine as long as I'm running the latest Omnisharp by setting `omnisharp.path` to `latest` in my settings.json. It doesn't seem that editing the Syntax.xml results in the generators rerunning though. @JoeRobich is there some known bug where additional files don't get updated in the Workspace?

**Comment by RikkiGibson** [comment]
> 
> 
> I want to make sure that we don't merge this before we've validated the experience works on vscode, or this will effectively break my ability to work.

I also work primarily in vscode and the ability to go-to-def etc. will be a significant inconvenience for me.

In addition, I'm concerned about moving forward to a model where, because the generated syntax isn't version controlled, it's difficult for us to audit changes to it over time, or for example review a change to the generator in terms of the diff of the output code.

**Comment by CyrusNajmabadi** [comment]
> In addition, I'm concerned about moving forward to a model where

These are fundamental development concerns.  I think if we're going to be pushing SGs as a solution to codegen, we'll need answers there.  If we can't even use them for a basic case like this for us (i.e. simple model produces simple codegen'ed api), i hesitate about us stating that others should use it in production settings.

**Comment by RikkiGibson** [comment]
Thinking over it a bit. I think that it's just a matter of distinguishing the "generator user" versus "generator author" experiences. We are the latter. In our case, the output of the generator when we feed in Syntax.xml for example is a test baseline. CI needs to fail when the baseline is out of sync with what the generator actually produces. So if we can make sure that happens, then I'm good.

**Comment by jasonmalinowski** [comment]
> We are the latter.

@RikkiGibson In this case, I think we're both, right? If you're adding new syntax nodes as part of feature development, do you you need to show the diff of the underlying generated file in that case? You might want to reference it while you're developing, but is that actually reviewed on it's own? I agree that if you're _changing_ the generator then it's a different story.

> In our case, the output of the generator when we feed in Syntax.xml for example is a test baseline. CI needs to fail when the baseline is out of sync with what the generator actually produces. So if we can make sure that happens, then I'm good.

@RikkiGibson can you clarify this a bit? The CI today fails if you failed to run the generator before checking in; the point of moving this to a source generator is it's not possible to be out of sync anymore. (indeed that's part of the point of generators in the first place...)

**Comment by RikkiGibson** [comment]
> In this case, I think we're both, right?

Yes.

> If you're adding new syntax nodes as part of feature development, do you you need to show the diff of the underlying generated file in that case?

No, that is not particularly useful to do. Perhaps if an input change exposed a bug in the generator, but that seems like something you'd examine locally, not in code review.

> can you clarify this a bit?

Sorry, I stated this too much in terms of "how it works right now". When someone PRs an implementation change for a generator like CSharpSyntaxGenerator I just want to be able to easily review the diff of the generator output.

Less importantly: I can also conceive of situations where I'd want to f12 on a generated method, then run a blame to figure out when the generator was changed to produce the implementation that particular way.

**Comment by CyrusNajmabadi** [comment]
> Less importantly: I can also conceive of situations where I'd want to f12 on a generated method, then run a blame to figure out when the generator was changed to produce the implementation that particular way.

This is also in teh scope of "i think everyone who uses source generators would want this".  So by dogfooding this, we get a realistic understanding around requests for that sort of thing that will likely come in the future. :)

**Comment by chsienki** [comment]
@jasonmalinowski which analyzer rules are failing to be taken into account? I see a couple in the correctness leg, but they look like they're in actual files, not generated ones?

**Comment by sharwell** [comment]
While I like the concept of this change, I'm opposed to adopting this strategy prior to #48123 on the grounds that it would be too disruptive to normal workflows with no available alternative.

**Comment by CyrusNajmabadi** [comment]
> I'm opposed to adopting this strategy prior to #48123 on the grounds that it would be too disruptive to normal workflows with no available alternative.

I think this is the point.  We will not be motivated to ensure SG scenarios are a good enough experience unless we ahve to deal with the pain of what the SG experience is in a mainline scenario for us.  If we can't dogfood this for precisely teh sorts of scenarios SGs are for, then it's not ready for customers to be using for those same sorts of scenarios.

**Comment by 333fred** [comment]
For those who don't know the fancy words, what are richnav/source browser? Is this like http://sourceroslyn.io?

**Comment by sharwell** [comment]
@333fred Source Browser is the tool behind sourceroslyn.io. Rich Navigation / LSIF are new technology to power tools like this, with a goal of native GitHub and Code Spaces integration.

**Comment by davidwengier** [comment]
I might be alone in this, but I personally think of code generated by source generators to be internal implementation details, much like the backing fields of auto-props or the async state machines that are generated by the compiler. I wouldn't want to see them in Source Browser or other sources that don't use IL decompilation. Perhaps this scenario is not a good fit for source generators, and we should look for others to dogfood?

**Comment by jasonmalinowski** [comment]
@davidwengier: I'm confused which way you're arguing. Are you arguing:

1. You wouldn't expect to see the code generated by this generator, which means it's a good generator and the concerns about sourceroslyn.io aren't applicable, or
2. You do want to look at this code, and thus we should not be using a generator for this.

**Comment by davidwengier** [comment]
@jasonmalinowski I'm not trying to make a judgement about _this_ generator specifically, but more the general case. I wouldn't expect to see the code generated by _any_ generator (1), so I guess therefore if people need to see the code then I don't think this is a good fit for a source generator (2).

**Comment by CyrusNajmabadi** [comment]
>  so I guess therefore if people need to see the code then I don't think this is a good fit for a source generator (2).

1. couldn't that apply to any SG?
2. why would we need to see the generated code for SyntaxGenerator?  it's literally mindnumbing :)

**Comment by davidwengier** [comment]
> couldn't that apply to any SG?
Yes, I think it does.
> why would we need to see the generated code for SyntaxGenerator? it's literally mindnumbing :)
I have no personal opinion on this, but based on the comments in this thread it seems that some people do. If nobody ever wanted to see generated source outside of debugging or go to def, then presumably #48123 wouldn't exist.

**Comment by jasonmalinowski** [comment]
So we're still adding plenty of features and ways for people to see the code; that's not going away. And there's always _somebody_ who potentially needs to look at the source for anything when you're debugging something. Perhaps thinking in general terms, I don't ever look at the IL generated by my C# code...until I thought I hit a compiler bug and had to chase it down. The generated syntax logic is absolutely mind-numbing as @CyrusNajmabadi mentions, but you might have to look at it if you're actually trying to figure out what happened, or the generator didn't generate what you might expect.

As far as sourceroslyn.io, I think the biggest goal there at least would be that find references and such isn't broken so people can still find references to certain syntax nodes or syntax nodes. Trying to load the generated file is an absolutely crappy experience today (it looks like it no longer hangs my browser at least), and so I do want to push for an experience there for what people actually _want_ as opposed to just

**Comment by jaredpar** [comment]
I think it's important as a *long term* goal that you be able to see the generated source in places like sourceroslyn.io. Yes I agree, it's a bad experience today due to the enormous size of the generated code. Further it is indeed mind numbing and not worth seeing on a day to day basis. 

There are times though when it is important. Particularly when you are changing the generator itself or trying to understand why certain nullable decisions were made (speaking from personal experience here).

My 2 cents is we should push forward with this change, not the experience gap and follow up on fixing it.

**Comment by davidwengier** [comment]
> until I thought I hit a compiler bug and had to chase it down

This brings up a related point, which is how do Source Generators play with symbol servers when debugging dumps etc. This was brought up on a PR for another source generator, in Winforms: https://github.com/dotnet/winforms/pull/4043#issuecomment-702401087

**Comment by chsienki** [comment]
The analyzer consistency problem wasn't directly caused by Jason's changes, but rather the changes are surfacing an issue we already had:

The net5.0 sdk now includes the NetAnalyzers package directly. We also reference it via NuGet, meaning we're trying to load the same analyzers twice. Previously when we disabled analyzers we did so by removing them at the MSBuild layer, meaning we never actually tried to load them. As part of this PR we're now disabling them via the new flag, so we're now loading them (even though they don't run) meaning we correctly detect the duplicate analyzers.

For now I've disabled the in-box SDK analyzers, as we want to run them on other TFM builds too (and keep them in-sync) but we might consider conditionally including/excluding depending on the TFM we're building in the future.

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-compiler could I get some reviews?

File: src/Tools/BuildBoss/ProjectCheckerUtil.cs
Comment (jaredpar) [reply src/Tools/BuildBoss/ProjectCheckerUtil.cs:146]: Think this is sensible

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (jaredpar) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:34]: Should we emit a diagnostic here? Seems like this would be an error scenario.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (jaredpar) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:41]: Think this diagnostic should be added or this should be converted to an issue before we merge.

Comment (jasonmalinowski) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs]: Updated.

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (jasonmalinowski) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:59]: Updated.

**Comment by jasonmalinowski** [comment]
@jaredpar @chsienki Diagnostic handling updated. Ready for review again.

**Review by jaredpar — APPROVED** [review]

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (chsienki) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:114]: @jasonmalinowski I'm 100% fine with adding at string builder overload. Do you want me to add it to this PR?

Comment (jasonmalinowski) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:114]: I'd say no, so if we have to revert this we're also not reverting a public API change that somebody else might depend on.

Comment (jasonmalinowski) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:114]: Filed https://github.com/dotnet/roslyn/issues/49082 to track it.

**Review by chsienki — APPROVED** [review]

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (Neme12) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:114]: @jasonmalinowski 
In #49689, the conclusion was that there shouldn't be a StringBuilder overload because it's mutable, but it seems like being able to create a SourceText from a TextReader like this has the exactly the same problem?

Comment (Neme12) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:114]: Actually no it doesn't, because `SourceText.From(TextReader)` calls `ReadToEnd` and allocates the string anyway, so this doesn't really avoid the allocation 😄
https://sourceroslyn.io/#Microsoft.CodeAnalysis/Text/SourceText.cs,121

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs
Comment (Neme12) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs:118]: Creating a `TextReader` from `SourceText` seems like this is something that should be provided by Roslyn as well. Otherwise, people are more likely to create a `string` out of it than make their own reader, which is less efficient.

**Pull request merged**
